File: python/src/openharness/runner.py
"""The agent loop.

    result = await run(agent, "What's the weather in Paris?")

Each turn: build the prompt, call the model, run any requested tools (in
parallel), feed results back, repeat until the model answers without tool
calls. Handoffs switch the active agent. Limits, cancellation and guardrails
are checked at every step.
"""

from __future__ import annotations

import asyncio
import inspect
import json
import time
from collections.abc import AsyncIterator, Awaitable, Callable
from typing import Any

from .agent import Agent, Handoff
from .context import CancelToken, RunContext, RunLimits
from .errors import (
    GuardrailTripped,
    MaxToolCallsExceeded,
    MaxTurnsExceeded,
    OpenHarnessError,
    OutputValidationError,
    RunCancelled,
    RunStopped,
    RunTimeout,
    TokenBudgetExceeded,
    ToolError,
)
from .guardrails import Guardrail
from .memory import memory_tools
from .models import Model
from .models.base import ModelEvent
from .result import RunResult, RunStream
from .schema import SchemaValidationError, output_schema_for, parse_output, schema_name
from .sessions import Session
from .tools import Tool
from .tracing import Trace, TraceProcessor, global_processors
from .types import Message, ModelRequest, ModelResponse, ModelSettings, OutputSchema, StreamEvent, ToolCall, to_json

Approver = Callable[[RunContext, ToolCall], "bool | Awaitable[bool]"]
EventHook = Callable[[StreamEvent], "None | Awaitable[None]"]


class _Deadline:
    def __init__(self, limits: RunLimits, cancel: CancelToken):
        self.limits = limits
        self.cancel = cancel
        self.start = time.monotonic()

    def remaining(self) -> float | None:
        if self.limits.timeout_seconds is None:
            return None
        return self.limits.timeout_seconds - (time.monotonic() - self.start)

    async def guard(self, awaitable: Awaitable[Any]) -> Any:
        """Await something, aborting on cancellation or when the run's time is up."""
        task = asyncio.ensure_future(awaitable)
        stopper = asyncio.ensure_future(self.cancel.wait())
        remaining = self.remaining()
        try:
            done, _ = await asyncio.wait(
                {task, stopper},
                timeout=None if remaining is None else max(remaining, 0),
                return_when=asyncio.FIRST_COMPLETED,
            )
        finally:
            stopper.cancel()
        if task in done:
            return task.result()
        task.cancel()
        try:
            await task
        except (asyncio.CancelledError, Exception):
            pass
        if self.cancel.cancelled:
            raise RunCancelled(f"Run cancelled: {self.cancel.reason}")
        raise RunTimeout(f"Run exceeded timeout of {self.limits.timeout_seconds}s")


async def _maybe_await(value: Any) -> Any:
    return await value if inspect.isawaitable(value) else value


def _input_messages(value: str | Message | list[Message]) -> list[Message]:
    if isinstance(value, str):
        return [Message.user(value)]
    if isinstance(value, Message):
        return [value]
    return list(value)


async def _run_guardrails(guardrails: list[Guardrail], ctx: RunContext, value: Any, stage: str, trace: Trace) -> Any:
    for g in guardrails:
        with trace.span("guardrail", g.name, stage=stage) as sp:
            res = await g.check(ctx, value)
            sp.set(tripped=res.tripwire)
        if res.tripwire:
            raise GuardrailTripped(f"{stage} guardrail '{g.name}' tripped", g.name, stage, res.info)
        if res.replacement is not None:
            value = res.replacement
    return value


async def _stream_loop(
    agent: Agent,
    input: str | Message | list[Message],
    *,
    session: Session | None = None,
    deps: Any = None,
    model: str | Model | None = None,
    limits: RunLimits | None = None,
    model_settings: ModelSettings | None = None,
    cancel_token: CancelToken | None = None,
    approve: Approver | None = None,
    trace: TraceProcessor | list[TraceProcessor] | None = None,
    trace_metadata: dict[str, Any] | None = None,
    streaming: bool = True,
    _parent_trace: Trace | None = None,
) -> AsyncIterator[StreamEvent]:
    limits = limits or agent.limits or RunLimits()
    cancel = cancel_token or CancelToken()
    deadline = _Deadline(limits, cancel)
    procs: list[TraceProcessor] = trace if isinstance(trace, list) else ([trace] if trace else [])
    tr = _parent_trace or Trace(agent.name, [*procs, *global_processors()], trace_metadata)
    owns_trace = _parent_trace is None

    history = await session.get_messages() if session else []
    inputs = _input_messages(input)
    ctx: RunContext = RunContext(agent=agent, deps=deps, trace=tr, cancel_token=cancel)
    new: list[Message] = []
    current = agent
    output_retries = 0
    mcp_cache: dict[int, list[Tool]] = {}

    def partial(text: str = "") -> RunResult:
        return RunResult(None, text, new, current, ctx.usage, ctx.turn, ctx.tool_calls, tr, history)

    def event(type_: str, **data: Any) -> StreamEvent:
        return StreamEvent(type_, data, agent=current.name)

    run_span_cm = tr.span("run" if owns_trace else "agent_tool", agent.name)
    run_span = run_span_cm.__enter__()
    try:
        # ---- input guardrails (on the latest user text)
        if agent.input_guardrails:
            idx = next((i for i in range(len(inputs) - 1, -1, -1) if inputs[i].role == "user"), None)
            if idx is not None:
                replaced = await deadline.guard(
                    _run_guardrails(agent.input_guardrails, ctx, inputs[idx].content, "input", tr)
                )
                if replaced != inputs[idx].content:
                    inputs[idx] = Message.user(str(replaced))
        new.extend(inputs)
        yield event("agent_start", agent=current.name)

        while True:
            # ---- limits
            if cancel.cancelled:
                raise RunCancelled(f"Run cancelled: {cancel.reason}")
            if limits.max_turns is not None and ctx.turn >= limits.max_turns:
                raise MaxTurnsExceeded(f"Agent stopped after reaching max_turns={limits.max_turns}")
            if limits.max_total_tokens is not None and ctx.usage.total_tokens >= limits.max_total_tokens:
                raise TokenBudgetExceeded(
                    f"Token budget of {limits.max_total_tokens} used up ({ctx.usage.total_tokens} tokens)"
                )
            rem = deadline.remaining()
            if rem is not None and rem <= 0:
                raise RunTimeout(f"Run exceeded timeout of {limits.timeout_seconds}s")
            ctx.turn += 1
            ctx.agent = current

            # ---- assemble tools
            tools: list[Tool] = list(current.tools)
            if current.memory is not None:
                tools += memory_tools(current.memory)
            for server in current.mcp_servers:
                key = id(server)
                if key not in mcp_cache:
                    with tr.span("mcp", server.name, action="list_tools"):
                        mcp_cache[key] = await deadline.guard(server.list_tools())
                tools += mcp_cache[key]
            handoffs = {h.name: h for h in current.handoff_objects()}
            tools += [h.tool() for h in handoffs.values()]
            by_name: dict[str, Tool] = {}
            for t in tools:
                by_name.setdefault(t.name, t)

            # ---- system prompt
            system = await deadline.guard(_maybe_await(current.render_instructions(ctx)))
            if current.memory is not None:
                query = next((m.content for m in reversed([*history, *new]) if m.role == "user"), "")
                found = await current.memory.search(query, limit=5)
                if found:
                    system += "\n\nRelevant long-term memories:\n" + "\n".join(f"- {m.text}" for m in found)

            m = current.get_model(model)
            settings = current.model_settings.merged(model_settings)
            out_schema = None
            if current.output_type is not None:
                schema = output_schema_for(current.output_type)
                native = not (m.provider == "gemini" and by_name)  # Gemini: schema + tools not combined
                if native:
                    out_schema = OutputSchema(schema_name(current.output_type), schema)
                else:
                    system += (
                        "\n\nWhen you give your final answer, reply with only a JSON object that matches "
                        f"this JSON Schema:\n{json.dumps(schema)}"
                    )

            request = ModelRequest(
                system=system,
                messages=[*history, *new],
                tools=[t.spec() for t in by_name.values()],
                output_schema=out_schema,
                settings=settings,
            )

            # ---- call the model
            response: ModelResponse | None = None
            with tr.span("model", f"{m.provider}:{m.model}", agent=current.name, turn=ctx.turn) as msp:
                if streaming:
                    gen = m.stream(request).__aiter__()
                    while True:
                        try:
                            ev: ModelEvent = await deadline.guard(gen.__anext__())
                        except StopAsyncIteration:
                            break
                        if ev.type == "text_delta":
                            yield event("text_delta", delta=ev.data["delta"])
                        elif ev.type == "done":
                            response = ev.response
                    if response is None:
                        raise OpenHarnessError(f"{m!r} stream ended without a response")
                else:
                    response = await deadline.guard(m.generate(request))
                msp.set(
                    usage=response.usage.to_dict(),
                    stop_reason=response.stop_reason,
                    tool_calls=[tc.name for tc in response.message.tool_calls],
                )
            ctx.usage.add(response.usage)
            msg = response.message
            msg.name = current.name
            new.append(msg)
            ctx.messages = [*history, *new]
            yield event("message", message=msg)

            # ---- tool calls
            if msg.tool_calls:
                calls = msg.tool_calls
                if limits.max_tool_calls is not None and ctx.tool_calls + len(calls) > limits.max_tool_calls:
                    new.pop()  # keep history valid: drop the unanswered tool request
                    raise MaxToolCallsExceeded(f"Agent stopped after reaching max_tool_calls={limits.max_tool_calls}")
                ctx.tool_calls += len(calls)
                handoff_target: Handoff | None = None
                for tc in calls:
                    yield event("tool_call", id=tc.id, name=tc.name, arguments=tc.arguments)

                async def execute(tc: ToolCall, by_name: dict[str, Tool] = by_name) -> Message:
                    nonlocal handoff_target
                    t = by_name.get(tc.name)
                    if t is None:
                        return Message.tool(
                            tc.id,
                            tc.name,
                            f"Error: unknown tool '{tc.name}'. Available: {', '.join(by_name) or 'none'}",
                            is_error=True,
                        )
                    h = t.metadata.get("handoff")
                    if h is not None:
                        if handoff_target is None:
                            handoff_target = h
                            return Message.tool(tc.id, tc.name, f"Transferred to {h.agent.name}.")
                        return Message.tool(tc.id, tc.name, "Ignored: another handoff was already made.", is_error=True)
                    with tr.span("tool", tc.name, parent=msp, arguments=tc.arguments) as tsp:
                        if t.requires_approval(tc.arguments):
                            ok = await _maybe_await(approve(ctx, tc)) if approve else False
                            tsp.set(approved=ok)
                            if not ok:
                                return Message.tool(
                                    tc.id, tc.name, "Error: the user did not approve this tool call.", is_error=True
                                )
                        try:
                            value = await t.invoke(tc.arguments, ctx)
                            out = to_json(value)
                            tsp.set(output=out[:2000])
                            return Message.tool(tc.id, tc.name, out)
                        except (RunStopped, asyncio.CancelledError):
                            raise
                        except ToolError as e:
                            tsp.error = str(e)
                            return Message.tool(tc.id, tc.name, f"Error: {e}", is_error=True)
                        except Exception as e:  # tool bugs go back to the model, not up the stack
                            tsp.error = f"{type(e).__name__}: {e}"
                            return Message.tool(tc.id, tc.name, f"Error: {type(e).__name__}: {e}", is_error=True)

                results = await deadline.guard(asyncio.gather(*(execute(tc) for tc in calls)))
                for r in results:
                    new.append(r)
                    yield event("tool_result", id=r.tool_call_id, name=r.name, output=r.content, is_error=r.is_error)
                if handoff_target is not None:
                    prev = current
                    with tr.span("handoff", f"{prev.name} -> {handoff_target.agent.name}"):
                        current = handoff_target.agent
                        if handoff_target.input_filter:
                            kept = handoff_target.input_filter([*history, *new])
                            history, new = [], list(kept)
                    yield event("handoff", **{"from": prev.name, "to": current.name})
                    yield event("agent_start", agent=current.name)
                continue

            # ---- final answer
            text = msg.content
            output: Any = text
            if current.output_type is not None:
                try:
                    output = parse_output(text, current.output_type, output_schema_for(current.output_type))
                except (SchemaValidationError, ValueError, TypeError) as e:
                    if output_retries >= limits.max_output_retries:
                        raise OutputValidationError(f"Structured output failed validation: {e}") from e
                    output_retries += 1
                    new.append(
                        Message.user(
                            f"Your reply did not match the required JSON schema: {e}. "
                            "Reply again with only the corrected JSON."
                        )
                    )
                    continue
            if current.output_guardrails:
                output = await deadline.guard(_run_guardrails(current.output_guardrails, ctx, output, "output", tr))
                if isinstance(output, str):
                    text = output
            result = RunResult(output, text, new, current, ctx.usage, ctx.turn, ctx.tool_calls, tr, history)
            if session is not None:
                await session.add_messages(new)
            run_span.set(turns=ctx.turn, usage=ctx.usage.to_dict(), last_agent=current.name)
            yield event("run_end", result=result)
            return
    except RunStopped as e:
        if e.partial is None:
            e.partial = partial()
        run_span.error = f"{type(e).__name__}: {e}"
        raise
    finally:
        run_span_cm.__exit__(None, None, None)
        if owns_trace:
            tr.finish()


async def run(agent: Agent, input: str | Message | list[Message], **kwargs: Any) -> RunResult:
    """Run an agent to completion and return the result.

    Keyword args: session, deps, model, limits, model_settings, cancel_token,
    approve, trace, trace_metadata, on_event.
    """
    on_event: EventHook | None = kwargs.pop("on_event", None)
    result: RunResult | None = None
    async for ev in _stream_loop(agent, input, streaming=on_event is not None, **kwargs):
        if on_event is not None:
            await _maybe_await(on_event(ev))
        if ev.type == "run_end":
            result = ev.data["result"]
    assert result is not None
    return result


def run_stream(agent: Agent, input: str | Message | list[Message], **kwargs: Any) -> RunStream:
    """Start a run and stream events. See ``RunStream``."""
    token = kwargs.pop("cancel_token", None) or CancelToken()
    return RunStream(_stream_loop(agent, input, cancel_token=token, streaming=True, **kwargs), token)


def run_sync(agent: Agent, input: str | Message | list[Message], **kwargs: Any) -> RunResult:
    """Blocking version of ``run`` for scripts and notebooks without an event loop."""
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(run(agent, input, **kwargs))
    raise OpenHarnessError("run_sync() cannot be used inside a running event loop; use 'await run(...)'")

File: typescript/src/runner.ts
/**
 * The agent loop.
 *
 *   const result = await run(agent, "What's the weather in Paris?");
 *
 * Each turn: build the prompt, call the model, run any requested tools (in
 * parallel), feed results back, repeat until the model answers without tool
 * calls. Handoffs switch the active agent. Limits, cancellation and
 * guardrails are checked at every step.
 */

import type { Agent, Handoff } from "./agent.js";
import { CancelToken, DEFAULT_LIMITS, RunContext, type RunLimits } from "./context.js";
import {
  GuardrailTripped,
  MaxToolCallsExceeded,
  MaxTurnsExceeded,
  OpenHarnessError,
  OutputValidationError,
  RunCancelled,
  RunStopped,
  RunTimeout,
  TokenBudgetExceeded,
  ToolError,
} from "./errors.js";
import { runGuardrail, type Guardrail } from "./guardrails.js";
import { memoryTools } from "./memory.js";
import type { Model } from "./models/index.js";
import { RunStream, type RunResult } from "./result.js";
import { extractJson, outputSchemaFor, SchemaValidationError, validate } from "./schema.js";
import type { Session } from "./sessions.js";
import { invokeTool, requiresApproval, toolSpec, type Tool } from "./tools.js";
import { globalProcessors, Trace, type Span, type TraceProcessor } from "./tracing.js";
import {
  mergeSettings,
  Message,
  toJSONText,
  type ModelRequest,
  type ModelResponse,
  type ModelSettings,
  type StreamEvent,
  type ToolCall,
} from "./types.js";

export type Approver = (ctx: RunContext<any>, call: ToolCall) => boolean | Promise<boolean>;

export interface RunOptions<TDeps = unknown> {
  session?: Session;
  /** Your own objects (db handles, current user). Passed to tools; never sent to the model. */
  deps?: TDeps;
  /** Override the agent's model for this run. */
  model?: string | Model;
  limits?: RunLimits;
  modelSettings?: ModelSettings;
  /** Standard AbortSignal; aborting it cancels the run. */
  signal?: AbortSignal;
  cancelToken?: CancelToken;
  /** Decides whether tools marked `needsApproval` may run. Without it they are denied. */
  approve?: Approver;
  trace?: TraceProcessor | TraceProcessor[];
  traceMetadata?: Record<string, unknown>;
  /** Called for every stream event (enables streaming from the provider). */
  onEvent?: (event: StreamEvent) => void | Promise<void>;
  /** @internal nested runs (agent as tool) share the parent trace */
  parentTrace?: Trace;
}

type Input = string | Message | Message[];

function inputMessages(input: Input): Message[] {
  if (typeof input === "string") return [Message.user(input)];
  return Array.isArray(input) ? [...input] : [input];
}

async function runGuardrails(list: Guardrail[], ctx: RunContext<any>, value: unknown, stage: "input" | "output", trace: Trace, parent: Span) {
  for (const g of list) {
    const res = await trace.span("guardrail", g.name, parent, { stage }, async (sp) => {
      const r = await runGuardrail(g, ctx, value);
      sp.set({ tripped: r.tripwire });
      return r;
    });
    if (res.tripwire) throw new GuardrailTripped(`${stage} guardrail '${g.name}' tripped`, g.name, stage, res.info);
    if (res.replacement !== undefined) value = res.replacement;
  }
  return value;
}

async function* loop<TOutput>(agent: Agent<TOutput>, input: Input, opts: RunOptions<any>, streaming: boolean): AsyncGenerator<StreamEvent> {
  const limits = { ...DEFAULT_LIMITS, ...(agent.limits ?? {}), ...(opts.limits ?? {}) };
  const token = opts.cancelToken ?? new CancelToken();
  const signals = [token.signal];
  if (opts.signal) signals.push(opts.signal);
  const timeout = limits.timeoutSeconds ? AbortSignal.timeout(limits.timeoutSeconds * 1000) : undefined;
  if (timeout) signals.push(timeout);
  const runSignal = AbortSignal.any(signals);

  const stopError = (): RunStopped => {
    if (token.cancelled) return new RunCancelled(`Run cancelled: ${token.reason}`);
    if (opts.signal?.aborted) return new RunCancelled(`Run cancelled: ${String(opts.signal.reason ?? "aborted")}`);
    return new RunTimeout(`Run exceeded timeout of ${limits.timeoutSeconds}s`);
  };
  const guard = <T>(p: Promise<T>): Promise<T> => {
    if (runSignal.aborted) {
      p.catch(() => undefined);
      return Promise.reject(stopError());
    }
    return new Promise<T>((resolve, reject) => {
      const onAbort = () => reject(stopError());
      runSignal.addEventListener("abort", onAbort, { once: true });
      p.then(
        (v) => {
          runSignal.removeEventListener("abort", onAbort);
          resolve(v);
        },
        (e) => {
          runSignal.removeEventListener("abort", onAbort);
          reject(runSignal.aborted ? stopError() : e);
        },
      );
    });
  };

  const procs = opts.trace ? (Array.isArray(opts.trace) ? opts.trace : [opts.trace]) : [];
  const ownsTrace = !opts.parentTrace;
  const trace = opts.parentTrace ?? new Trace(agent.name, [...procs, ...globalProcessors()], opts.traceMetadata);
  const ctx = new RunContext(agent, opts.deps, trace, runSignal);
  let history = opts.session ? await opts.session.getMessages() : [];
  const inputs = inputMessages(input);
  let fresh: Message[] = [];
  let current: Agent<any> = agent;
  let outputRetries = 0;
  const mcpCache = new Map<object, Tool[]>();

  const partial = (): RunResult<any> => ({
    output: undefined,
    text: "",
    newMessages: fresh,
    history,
    lastAgent: current,
    usage: ctx.usage,
    turns: ctx.turn,
    toolCalls: ctx.toolCalls,
    trace,
  });
  const ev = (type: StreamEvent["type"], data: Record<string, unknown>): StreamEvent => ({ type, data, agent: current.name, timestamp: Date.now() / 1000 });

  const runSpan = trace.startSpan(ownsTrace ? "run" : "agent_tool", agent.name);
  let failure: unknown;
  try {
    if (agent.inputGuardrails.length) {
      for (let i = inputs.length - 1; i >= 0; i--) {
        if (inputs[i].role !== "user") continue;
        const replaced = await guard(runGuardrails(agent.inputGuardrails, ctx, inputs[i].content, "input", trace, runSpan));
        if (replaced !== inputs[i].content) inputs[i] = Message.user(String(replaced));
        break;
      }
    }
    fresh.push(...inputs);
    yield ev("agent_start", { agent: current.name });

    for (;;) {
      if (runSignal.aborted) throw stopError();
      if (limits.maxTurns != null && ctx.turn >= limits.maxTurns) throw new MaxTurnsExceeded(`Agent stopped after reaching maxTurns=${limits.maxTurns}`);
      if (limits.maxTotalTokens != null && ctx.usage.totalTokens >= limits.maxTotalTokens)
        throw new TokenBudgetExceeded(`Token budget of ${limits.maxTotalTokens} used up (${ctx.usage.totalTokens} tokens)`);
      ctx.turn += 1;
      ctx.agent = current;

      // ---- tools
      const tools: Tool[] = [...current.tools];
      if (current.memory) tools.push(...memoryTools(current.memory));
      for (const server of current.mcpServers) {
        if (!mcpCache.has(server)) {
          const list = await trace.span("mcp", server.name, runSpan, { action: "listTools" }, () => guard(server.listTools()));
          mcpCache.set(server, list);
        }
        tools.push(...mcpCache.get(server)!);
      }
      const handoffs = current.handoffObjects();
      tools.push(...handoffs.map((h) => h.tool()));
      const byName = new Map<string, Tool>();
      for (const t of tools) if (!byName.has(t.name)) byName.set(t.name, t);

      // ---- system prompt
      let system = await guard(current.renderInstructions(ctx));
      if (current.memory) {
        const query = [...history, ...fresh].reverse().find((m) => m.role === "user")?.content ?? "";
        const found = await current.memory.search(query, 5);
        if (found.length) system += `\n\nRelevant long-term memories:\n${found.map((m) => `- ${m.text}`).join("\n")}`;
      }

      const model = current.getModel(opts.model);
      let outputSchema: ModelRequest["outputSchema"];
      if (current.outputType) {
        const { schema } = outputSchemaFor(current.outputType);
        const native = !(model.provider === "gemini" && byName.size > 0); // Gemini: schema + tools not combined
        if (native) outputSchema = { name: String(current.outputType.title ?? "output").replace(/[^a-zA-Z0-9_-]/g, "_"), schema };
        else system += `\n\nWhen you give your final answer, reply with only a JSON object that matches this JSON Schema:\n${JSON.stringify(schema)}`;
      }
      const request: ModelRequest = {
        system,
        messages: [...history, ...fresh],
        tools: [...byName.values()].map(toolSpec),
        outputSchema,
        settings: mergeSettings(current.modelSettings, opts.modelSettings),
        signal: runSignal,
      };

      // ---- model call
      const modelSpan = trace.startSpan("model", `${model.provider}:${model.model}`, runSpan, { agent: current.name, turn: ctx.turn });
      let response: ModelResponse | undefined;
      try {
        if (streaming) {
          const it = model.stream(request)[Symbol.asyncIterator]();
          for (;;) {
            const step = await guard(it.next());
            if (step.done) break;
            const e = step.value;
            if (e.type === "text_delta") yield ev("text_delta", { delta: e.delta });
            else if (e.type === "done") response = e.response;
          }
          if (!response) throw new OpenHarnessError(`${model.provider}:${model.model} stream ended without a response`);
        } else {
          response = await guard(model.generate(request));
        }
        modelSpan.set({ usage: response.usage.toJSON(), stopReason: response.stopReason, toolCalls: (response.message.toolCalls ?? []).map((t) => t.name) });
        trace.endSpan(modelSpan);
      } catch (e) {
        trace.endSpan(modelSpan, e);
        throw e;
      }
      ctx.usage.add(response.usage);
      const msg = response.message;
      msg.name = current.name;
      fresh.push(msg);
      ctx.messages = [...history, ...fresh];
      yield ev("message", { message: msg });

      // ---- tool calls
      const calls = msg.toolCalls ?? [];
      if (calls.length) {
        if (limits.maxToolCalls != null && ctx.toolCalls + calls.length > limits.maxToolCalls) {
          fresh.pop();
          throw new MaxToolCallsExceeded(`Agent stopped after reaching maxToolCalls=${limits.maxToolCalls}`);
        }
        ctx.toolCalls += calls.length;
        for (const c of calls) yield ev("tool_call", { id: c.id, name: c.name, arguments: c.arguments });
        let handoffTarget: Handoff | undefined;

        const execute = async (tc: ToolCall): Promise<Message> => {
          const t = byName.get(tc.name);
          if (!t) return Message.tool(tc.id, tc.name, `Error: unknown tool '${tc.name}'. Available: ${[...byName.keys()].join(", ") || "none"}`, true);
          const h = t.metadata?.handoff as Handoff | undefined;
          if (h) {
            if (handoffTarget) return Message.tool(tc.id, tc.name, "Ignored: another handoff was already made.", true);
            handoffTarget = h;
            return Message.tool(tc.id, tc.name, `Transferred to ${h.agent.name}.`);
          }
          const span = trace.startSpan("tool", tc.name, modelSpan, { arguments: tc.arguments });
          try {
            if (requiresApproval(t, tc.arguments)) {
              const ok = opts.approve ? await opts.approve(ctx, tc) : false;
              span.set({ approved: ok });
              if (!ok) return Message.tool(tc.id, tc.name, "Error: the user did not approve this tool call.", true);
            }
            const out = toJSONText(await invokeTool(t, tc.arguments, ctx));
            span.set({ output: out.slice(0, 2000) });
            return Message.tool(tc.id, tc.name, out);
          } catch (e) {
            if (e instanceof RunStopped) throw e;
            span.error = e instanceof Error ? `${e.name}: ${e.message}` : String(e);
            const text = e instanceof ToolError ? e.message : e instanceof Error ? `${e.name}: ${e.message}` : String(e);
            return Message.tool(tc.id, tc.name, `Error: ${text}`, true);
          } finally {
            trace.endSpan(span);
          }
        };

        const results = await guard(Promise.all(calls.map(execute)));
        for (const r of results) {
          fresh.push(r);
          yield ev("tool_result", { id: r.toolCallId, name: r.name, output: r.content, isError: Boolean(r.isError) });
        }
        if (handoffTarget) {
          const target: Handoff = handoffTarget;
          const prev = current;
          const hs = trace.startSpan("handoff", `${prev.name} -> ${target.agent.name}`, runSpan);
          current = target.agent;
          if (target.options.inputFilter) {
            fresh = target.options.inputFilter([...history, ...fresh]);
            history = [];
          }
          trace.endSpan(hs);
          yield ev("handoff", { from: prev.name, to: current.name });
          yield ev("agent_start", { agent: current.name });
        }
        continue;
      }

      // ---- final answer
      let text = msg.content;
      let output: unknown = text;
      if (current.outputType) {
        try {
          const { schema, wrapped } = outputSchemaFor(current.outputType);
          const data = extractJson(text);
          validate(data, schema);
          output = wrapped ? (data as any).value : data;
        } catch (e) {
          if (!(e instanceof SchemaValidationError)) throw e;
          if (outputRetries >= limits.maxOutputRetries) throw new OutputValidationError(`Structured output failed validation: ${e.message}`);
          outputRetries += 1;
          fresh.push(Message.user(`Your reply did not match the required JSON schema: ${e.message}. Reply again with only the corrected JSON.`));
          continue;
        }
      }
      if (current.outputGuardrails.length) {
        output = await guard(runGuardrails(current.outputGuardrails, ctx, output, "output", trace, runSpan));
        if (typeof output === "string") text = output;
      }
      const result: RunResult<any> = {
        output,
        text,
        newMessages: fresh,
        history,
        lastAgent: current,
        usage: ctx.usage,
        turns: ctx.turn,
        toolCalls: ctx.toolCalls,
        trace,
      };
      if (opts.session) await opts.session.addMessages(fresh);
      runSpan.set({ turns: ctx.turn, usage: ctx.usage.toJSON(), lastAgent: current.name });
      yield ev("run_end", { result });
      return;
    }
  } catch (e) {
    failure = e;
    if (e instanceof RunStopped && !e.partial) e.partial = partial();
    throw e;
  } finally {
    trace.endSpan(runSpan, failure);
    if (ownsTrace) trace.finish();
  }
}

/** Run an agent to completion and return the result. */
export async function run<TOutput = string, TDeps = unknown>(agent: Agent<TOutput>, input: Input, options: RunOptions<TDeps> = {}): Promise<RunResult<TOutput>> {
  let result: RunResult<TOutput> | undefined;
  for await (const e of loop(agent, input, options, Boolean(options.onEvent))) {
    if (options.onEvent) await options.onEvent(e);
    if (e.type === "run_end") result = e.data.result;
  }
  return result!;
}

/** Start a run and stream events. See `RunStream`. */
export function runStream<TOutput = string, TDeps = unknown>(agent: Agent<TOutput>, input: Input, options: RunOptions<TDeps> = {}): RunStream<TOutput> {
  const token = options.cancelToken ?? new CancelToken();
  return new RunStream<TOutput>(loop(agent, input, { ...options, cancelToken: token }, true), token);
}
